### PR TITLE
finite-wasm: add a nightly feature

### PR DESCRIPTION
**Pull request by nagisa**
It seems to have been fairly uneventful, except for one thing: we get different error outputs from wasmer2 vs near-vm.

I do believe this is not a protocol-affecting change though, as these error messages don't make it onto the chain?

Review note: best reviewed commit-by-commit. See individual descriptions.

### CODE REVIEW

**Comment by nagisa** [comment]
So, here's the issue with wasmer2 (and moreso wasmer0) – they have issues around stack limits, which is showing up as a different error or an invocation of UB.

File: integration-tests/src/tests/runtime/sanity_checks.rs [hunk 1]
```diff
@@ -231,22 +239,33 @@ fn test_sanity_used_gas() {
...
+    // Executing `used_gas` costs `base_cost` plus an instruction to execute the `call` itself.
+    // When executing `used_gas` twice within a metered block, the returned values should differ by
+    // that amount.
+    let runtime_config = node.client.read().unwrap().runtime_config.clone();
+    let base_cost = runtime_config.wasm_config.ext_costs.gas_cost(ExtCosts::base);
+    let op_cost = match runtime_config.wasm_config.limit_config.contract_prepare_version {
+        // Entering a new metered block, all of its instructions are paid upfront.
+        // Therefore, the difference across blocks must be larger than `base_cost`,
+        // given that the block contains other instructions besides the call of
+        // `used_gas`.
```
Comment (jakmeier) [inline integration-tests/src/tests/runtime/sanity_checks.rs:251]: This comment is very confusing to me. What does that have to do with the expected ops to be 0 for wasmer0 and wasmer2?

I see it was previously at the assertion further below. Should it be moved back there?

File: runtime/near-vm-runner/src/tests/compile_errors.rs [hunk 2]
```diff
@@ -338,7 +379,7 @@ fn test_limit_locals_global() {
             .make(),
         )
         .expect(expect![[r#"
-            VMOutcome: balance 4 storage_usage 12 return data None burnt gas 139269213 used gas 139269213
+            VMOutcome: balance 4 storage_usage 12 return data None burnt gas 13001413761 used gas 13001413761
```
Comment (jakmeier) [inline runtime/near-vm-runner/src/tests/compile_errors.rs]: Wow that's a lot of extra gas! Close to 100x more burnt.
I assume that's expected with the finite wasm changes? At least 15625 locals and the extra burning of 12 Ggas seems to check out if we charge the regular wasm op cost (0.82 Mgas) for each local.

**Review by jakmeier — APPROVED** [review]
Looks good to me. Thanks for the nicely reviewable commit-by-commit prep.

Yes, different errors are okay, as long as the gas usage is the same.

Not sure what's up with the CI failure (`test_limit_locals` panics)...

File: core/primitives/src/views.rs [hunk 3]
```diff
@@ -2783,6 +2783,8 @@ mod tests {
         use crate::runtime::config::RuntimeConfig;
         use crate::views::RuntimeConfigView;
 
+        // FIXME: This is snapshotting a config used for *tests*, rather than proper production
+        // configurations. That seems… subpar?
```
Comment (Ekleog-NEAR) [inline core/primitives/src/views.rs:2787]: Related: https://github.com/near/nearcore/issues/8202

**Review by Ekleog-NEAR — APPROVED** [review]
LGTM :)

File: integration-tests/src/tests/runtime/sanity_checks.rs (hunk 1 above)
Comment (nagisa) [reply integration-tests/src/tests/runtime/sanity_checks.rs:251]: I think this is the most appropriate place, given that this match block is responsible for implementing the differentiation between old vs new; I’ll see about expanding this.

**Comment by nagisa** [comment]
> Not sure what's up with the CI failure (test_limit_locals panics)...

Wasmer0 can't properly deal with stack overflow traps.

**Comment by nagisa** [comment]
One of the remaining failures is in a test that effectively runs a `loop {}` to termination (by gas exhaustion.)

This is actually fine, but it takes a long time to complete now that the new version of instrumentation no longer gets intrinsic-ified due to a different set of host function calls.

I'm not sure if it is worth fixing it much, but I do wonder why it is that we run a test like this at all :thinking:

**Comment by jakmeier** [comment]
> One of the remaining failures is in a test that effectively runs a loop {} to termination (by gas exhaustion.) This test is actually fine, but it takes a long time to complete now that the new version of instrumentation no longer gets intrinsic-ified due to a different set of host function calls. Then the test gets killed off by nextest.

That sounds like a worse performance hit than I expected. Will this be an issue? Or is this just an artifact of using finite-wasm without nearvm? Otherwise, could it be that we actually need to fix the lost intrinsics optimization before we can release this?

> I'm not sure if it is worth fixing it much, but I do wonder why it is that we run a test like this at all thinking

Well, it's absolutely crucial that we terminate on all WASM programs within a reasonable time. `loop {}` is one of those extreme cases that's quite likely to fail first and yet it's very easy to understand. So I would argue it's a perfect test.

**Comment by nagisa** [comment]
A missing piece of nuance to the message above: making these calls intrinsic-ified in wasmer2 runtime wouldn't be too onerous, but releasing the changes and upgrading wasmer2 and all the related ceremony is what would be the painful part.

And again, it would only help in our test suite.

**Comment by jakmeier** [comment]
Yeah, if this is only a test issue for a combination of features that is never going to hit production, I fully agree that it's not worth fixing. Let's just work around it in what ever way is the easiest, even if that means ignoring the test or something similar.

**Pull request merged**